[6.1.x] use etcd quorum reads for querying the operation plan

a-palchikov opened the pull request:
## Description
<!--Required. Provide high-level overview of what the change is for.-->
Use consistent (quorum) reads when querying the operation plan to avoid reading stale values.

## Type of change
<!--Required. Keep only those that apply.-->

* Bug fix (non-breaking change which fixes an issue)
* Internal change (not necessarily a bug fix or a new feature)
* This change has a user-facing impact

## Linked tickets and other PRs
<!--Required. Keep only those that apply.-->

<!--This PR addresses the following issues.-->
* Refs https://github.com/gravitational/gravity/issues/2140

## TODOs
<!--Required. Keep only those that apply and check them off as they get completed.-->

- [x] Self-review the change
- [x] Perform manual testing
- [ ] Address review feedback

## Implementation
<!--Optional. Add any relevant implementation details that might help the reviewers.-->

## Testing done
<!--Required. Explain what kind of testing these changes underwent.-->
This PR is under testing

a-palchikov committed:
Only refuse to use etcd if itis been explicitly disabled on a node (e.g.
during etcd upgrades) - otherwise use exponential backoff in case of
transient errors.

a-palchikov committed:
Reverse the etcd condition - check for disabled instead as enabled has many interpretations

a-palchikov committed:
Refine service disabled status detection

r0mant commented on lib/update/reconciler.go:
```diff
@@ -65,34 +67,31 @@ func (r *reconciler) ReconcilePlan(ctx context.Context, plan storage.OperationPl
 // syncChangelog will sync changelog entries from src to dst storage
-func (r *reconciler) syncChangelog(src storage.Backend, dst storage.Backend) error {
-	return trace.Wrap(SyncChangelog(src, dst, r.cluster, r.operationID))
+func (r *reconciler) syncChangelog(ctx context.Context, src storage.Backend, dst storage.Backend) error {
```
I'm wondering if retrying reconciliation would actually help. I don't recall seeing reconciliation errors in the logs so I wonder if the problem may actually be a lever deeper, with etcd itself returning stale data due to its leader election and other timeouts? Maybe it would be better to retry the pre-requisites check itself several times to account for situations where etcd needs some time to figure out its leader elections and start returning correct data under load?

a-palchikov replied:
Not sure how retrying the pre-requisites check can help mitigate a stale read? If we're seeing a stale value from etcd, then we will not be able to tell whether it is stale or up-to-date unless we request the read to be from the quorum.

a-palchikov committed:
Extend etcd backend implementation to enable setting etcd-specific
configuration for reads to be able to do consistent reads when querying
operation plan during upgrades.

a-palchikov commented on lib/fsm/format.go:
```diff
@@ -59,6 +59,7 @@ func FormatOperationPlanJSON(w io.Writer, plan storage.OperationPlan) error {
 func FormatOperationPlanText(w io.Writer, plan storage.OperationPlan) {
 	var t tabwriter.Writer
 	t.Init(w, 0, 10, 5, ' ', 0)
+	fmt.Fprintln(&t, "Operation: ", OperationString(plan))
```
When display operation plan, I always missed the details about the operation itself.

a-palchikov commented on lib/update/reconciler.go:
```diff
@@ -84,22 +87,28 @@ func (r *reconciler) trySyncChangelogFromEtcd(ctx context.Context) error {
 // syncChangelog will sync changelog entries from src to dst storage
-func (r *reconciler) syncChangelog(src storage.Backend, dst storage.Backend) error {
-	return trace.Wrap(SyncChangelog(src, dst, r.cluster, r.operationID))
+func (r *reconciler) syncChangelog(ctx context.Context, src, dst storage.Backend) error {
```
I've wrapped the query in a retry loop but left the original precondition that skips the reads if the cluster is not healthy. 

Initially I tried to work it around by first introducing a stricter check for when to skip the reads by testing for the etcd service being masked (which is the case once it has been shutdown on the node) but it does not work in a multi-node setting until all etcd instances have been shut down.
I then attempted to explicitly skip plan syncs by configuring it as an attribute of the plan step and checking it in the engine but it required a lot of changes and I failed to come up with a way to pair it with the FSM as it would require FSM to know about the reconciler and be able to provide custom handling during upgrades.
In the end, I figured it was all not worth the effort and the changes.

a-palchikov commented on tool/gravity/cli/rpcagent.go:
```diff
@@ -544,6 +545,28 @@ func getGravityPackage() loc.Locator {
 	}
 }
 
+func (r deployAgentsRequest) String() string {
```
I added this to avoid the request logging binary data so we only log the relevant parts.

a-palchikov committed:
Actually use the quorum reads in the reconciler

r0mant commented on lib/storage/keyval/etcd.go:
```diff
@@ -101,6 +105,30 @@ func LocalEtcdConfig(retryTimeout time.Duration) (*ETCDConfig, error) {
+// e.g. return the latest committed value applied in a quorum of members
+func WithReadQuorum(quorum bool) EtcdOption {
+	return func(config *etcdOptions) {
+		logrus.WithField("quorum", quorum).Info("Specify quorum reads.")
```
Looks more like a Debug message than Info?

r0mant commented on lib/storage/keyval/leader.go:
```diff
@@ -48,7 +50,17 @@ func (b *electingBackend) StepDown() {
 	b.Leader.StepDown()
 }
 
+func (b *electingBackend) CopyWithOptions(opts ...EtcdOption) storage.Backend {
```
Nit: godoc.

r0mant commented on lib/update/cluster/engine.go:
```diff
@@ -316,8 +316,9 @@ func (f *engine) reconcilePlan(ctx context.Context) error {
+	type s string
 	fsm.FormatOperationPlanText(&buf, f.plan)
-	f.Debugf("Reconciled plan: %v.", buf.String())
+	f.WithField("plan", s(buf.String())).Debug("Reconciled plan.")
```
Why is the type conversion needed here?

a-palchikov replied:
Ah, I've been meaning to comment that - will do. Basically, it is an ugly workaround to avoid plan output being escaped when going through the text log formatter.

r0mant commented on lib/update/reconciler.go:
```diff
@@ -84,22 +87,29 @@ func (r *reconciler) trySyncChangelogFromEtcd(ctx context.Context) error {
-	return trace.Wrap(SyncChangelog(src, dst, r.cluster, r.operationID))
+func (r *reconciler) syncChangelog(ctx context.Context, src, dst storage.Backend) error {
+	b := backoff.NewExponentialBackOff()
+	b.MaxInterval = 1 * time.Minute
```
Maybe reduce max interval to potentially get more retry attempts?

r0mant commented on lib/update/reconciler.go:
```diff
@@ -84,22 +87,29 @@ func (r *reconciler) trySyncChangelogFromEtcd(ctx context.Context) error {
 
 	if shouldSync {
-		return trace.Wrap(r.syncChangelog(r.backend, r.localBackend))
+		consistentSrc := r.backend.CopyWithOptions(keyval.WithReadQuorum(true))
```
I would add a comment explaining why we're doing this here.

r0mant commented on lib/storage/keyval/leader.go:
```diff
@@ -48,7 +50,17 @@ func (b *electingBackend) StepDown() {
+	return &electingBackend{
+		Backend: b.backend,
+		Leader:  b.Leader,
+		backend: b.backend,
```
Hmm, what's the second "backend" parameter here for? Looks like everywhere it's used it's always set to the same value as "Backend"? Or is that just to have access to the actual type, not the interface?

a-palchikov replied:
Sure, I'll comment on that in code as well. The second copy of the backend so it can be manipulated since it needs access to the concrete value. I've kept the `Backend` as well since it's required for `electingBackend` to implicitly satisfy the `storage.Backend` interface.

a-palchikov committed:
Address review comments

a-palchikov committed:
Merge branch 'version/6.1.x' into dmitri/6.1.x/2140-etcd-retry
